aya/src/programs/uprobe.rs

```diff
@@ -450,3 +432,277 @@ fn resolve_symbol(path: &str, symbol: &str) -> Result<u64, ResolveSymbolError> {
         Ok(sym.address() - section.address() + offset)
     }
 }
+
+/// Error reading from /proc/pid/maps.
+#[derive(Debug, Error)]
+pub enum ProcMapError {
+    /// Unable to read /proc/pid/maps.
+    #[error(transparent)]
+    Read(io::Error),
+
+    /// Error parsing an integer.
+    #[error(transparent)]
+    ParseInt(#[from] ParseIntError),
+
+    /// Error parsing a line of /proc/pid/maps.
+    #[error("proc map entry parse error")]
+    Parse,
+}
+
+/// The memory maps of a process.
+///
+/// This is read from /proc/`pid`/maps.
+///
+/// The information here may be used to resolve addresses to paths.
+struct ProcMap {
+    // This is going to be used by USDT probes to resolve virtual addresses to
+    // library paths.
+    _entries: Vec<ProcMapEntry>,
+    libraries: HashSet<PathBuf>,
+}
+
+impl ProcMap {
+    /// Create a new [`ProcMap`] from a given pid.
+    fn new(pid: pid_t) -> Result<Self, ProcMapError> {
+        let maps_file = format!("/proc/{}/maps", pid);
+        let data = fs::read_to_string(maps_file).map_err(ProcMapError::Read)?;
+        Self::from_str(&data)
+    }
+
+    // Find the full path of a library by its name.
+    //
+    // This isn't part of the public API since it's really only useful for
+    // attaching uprobes.
+    fn find_library_path_by_name(&self, lib: &str) -> Result<Option<&str>, io::Error> {
+        let ret = if lib.contains(".so") {
+            self.libraries.iter().find(|p| {
+                p.file_name()
+                    .and_then(|s| s.to_str())
+                    .map(|s| s.starts_with(lib))
+                    .unwrap_or_default()
+            })
+        } else {
+            self.libraries.iter().find(|p| {
+                p.file_name()
+                    .and_then(|s| s.to_str())
+                    .map(|s| {
+                        s.strip_prefix(lib)
+                            .map(|k| k.starts_with(".so") || k.starts_with('-'))
+                            .unwrap_or_default()
+                    })
+                    .unwrap_or_default()
+            })
+        };
+
+        Ok(ret.map(|p| {
+            p.as_os_str()
+                .to_str()
+                .expect("library path should be valid UTF-8")
+        }))
+    }
+}
+
+impl FromStr for ProcMap {
```

Comment:
It is not a valid assumption that the /proc/$pid/map is UTF-8, since the library name can be non-UTF-8 (though this is very unlikely, it still is a potential use case).